lib/Db/ShareMapper.php

```diff
@@ -139,6 +139,27 @@ public function findAllSharesForNodeFor(string $nodeType, int $nodeId, string $r
 		return $this->findEntities($qb);
 	}
 
+	public function findAllSharesForNodeTo(string $nodeType, int $nodeId, string $receivingUser, array $receivingGroups): array {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->table)
+			->where(
+				$qb->expr()->orX(
+					$qb->expr()->andX(
+						$qb->expr()->eq('receiver', $qb->createNamedParameter($receivingUser, IQueryBuilder::PARAM_STR)),
+						$qb->expr()->eq('receiver_type', $qb->createNamedParameter('user',IQueryBuilder::PARAM_STR)),
+					),
+					$qb->expr()->andX(
+						$qb->expr()->in('receiver', $qb->createNamedParameter($receivingGroups, IQueryBuilder::PARAM_STR_ARRAY)),
+						$qb->expr()->eq('receiver_type', $qb->createNamedParameter('group',IQueryBuilder::PARAM_STR)),
+					)
+				)
+			)
+			->andWhere($qb->expr()->eq('node_type', $qb->createNamedParameter($nodeType, IQueryBuilder::PARAM_STR)))
```

Comment:
While at it please add an index as well for this one if it is worth, seems we don't have any on oc_tables_share yet 🙈

Reply:
That's the plan. We get a anonymized production dump, when experimenting with indexes and dev data, we always ended up with a full table scan either way.